src/frontegg-auth/src/auth.rs

```diff
@@ -314,206 +274,314 @@ impl Authentication {
+                    let recent_drop = last_drop
+                        .map(|dropped_at| dropped_at.elapsed() <= Duration::from_secs(drop_window))
+                        .unwrap_or(false);
+                    if receiver_count == 0 && !recent_drop {
+                        tracing::debug!(
+                            ?last_drop,
+                            ?password.client_id,
+                            "all listeners have dropped and none of them were recent!"
+                        );
+                        break;
+                    }
 
-            // If we failed to refresh, then close the channel.
-            let current_response = match current_response {
-                Ok(resp) => {
-                    tracing::debug!(?id, "refresh successful");
-                    resp
+                    // TODO(parkmycar): Allocating the labels here feels bad.
+                    inner
+                        .metrics
+                        .session_refresh_count
+                        .with_label_values(&[&receiver_count.to_string(), &recent_drop.to_string()])
+                        .inc();
+                    tracing::debug!(
+                        receiver_count,
+                        ?last_drop,
+                        ?password.client_id,
+                        "refreshing due to interest in the session"
+                    );
+
+                    // We still have interest, attempt to refresh the session.
+                    let res = inner.exchange_app_password(&expected_email, password).await;
+                    claims = match res {
+                        Ok(claims) => {
+                            tracing::debug!("refresh successful");
+                            claims
+                        }
+                        Err(e) => {
+                            tracing::warn!(error = ?e, "refresh failed");
+                            break;
+                        }
+                    };
+                    external_metadata_tx.send_replace(ExternalUserMetadata {
+                        admin: claims.is_admin,
+                        user_id: claims.user_id,
+                    });
                 }
-                Err(err) => {
-                    tracing::warn!(?id, ?err, "failed to refresh");
 
-                    // Dropping the channel will close it.
-                    drop(refresh_tx);
-                    break;
+                // The session has expired. Clean up the state.
+                {
+                    let mut sessions = inner.active_sessions.lock().expect("lock poisoned");
+                    sessions.remove(&password);
                 }
-            };
 
-            latest_response = current_response;
+                tracing::debug!(?password.client_id, "shutting down refresh task");
+                gauge.dec();
+            }
+            .instrument(debug_span!("frontegg_auth_refresh_task", client_id = %password.client_id))
```

Comment:
How useful is this span? It can be pretty long-lived, and so won't show up (right?) in tempo for steady state sessions that are refreshing. Should we instead instrument the inner auth call?

Reply:
You're right, this span wouldn't close until the refresh task did, which could theoretically be multiple days. Removed this span and instrumented `exchange_app_password`.